core/node/rules/can_add_event.go

```diff
@@ -540,9 +586,209 @@ func (params *aeParams) creatorIsMember() (bool, error) {
+	if err != nil {
+		return false, err
+	}
+	if hasTransaction {
+		// this is a derived event, so we don't return an error so that the user
+		// can retry adding the original event until it succeeds
+		return false, nil
+	}
 	return true, nil
 }
 
+func (ru *aeReceivedBlockchainTransactionRules) validReceivedBlockchainTransaction_IsUnique() (bool, error) {
+	// loop over all events in the view, check if the transaction is already in the view
+	userStreamView := ru.params.streamView.(events.UserStreamView)
+
+	hasTransaction, err := userStreamView.HasTransaction(ru.receivedTransaction.Transaction.GetReceipt())
+	if err != nil {
+		return false, err
+	}
+	if hasTransaction {
+		// this is a derived event, so we don't return an error so that the user
+		// can retry adding the original event until it succeeds
+		return false, nil
+	}
+	return true, nil
+}
+
+func (ru *aeBlockchainTransactionRules) validBlockchainTransaction_IsUnique() (bool, error) {
+	// loop over all events in the view, check if the transaction is already in the view
+	userStreamView := ru.params.streamView.(events.UserStreamView)
+
+	hasTransaction, err := userStreamView.HasTransaction(ru.transaction.GetReceipt())
+	if err != nil {
+		return false, err
+	}
+	if hasTransaction {
+		return false, RiverError(
+			Err_INVALID_ARGUMENT,
+			"transaction already exists in stream",
+			"streamId",
+			ru.params.streamView.StreamId(),
+			"transactionHash",
+			ru.transaction.GetReceipt().TransactionHash,
+		)
+	}
+	return true, nil
+}
+
+func (ru *aeBlockchainTransactionRules) validBlockchainTransaction_ReceiptMetadata() (bool, error) {
+	// check creator
+	switch ru.transaction.Kind {
+	case BlockchainTransactionKind_BLOCKCHAIN_TRANSACTION_KIND_UNSPECIFIED:
+		// for unspecified types, we don't need to check anything specific
+		// the other checks should make sure the transaction is valid and from this user
+		return true, nil
+	case BlockchainTransactionKind_BLOCKCHAIN_TRANSACTION_KIND_TIP:
+		// todo
+		return true, nil
+	default:
+		return false, RiverError(
+			Err_INVALID_ARGUMENT,
+			"unknown transaction type",
+			"transactionType",
+			ru.transaction.Kind,
+		)
+	}
+}
+
+func (ru *aeReceivedBlockchainTransactionRules) parentEventForReceivedBlockchainTransaction() (*DerivedEvent, error) {
+	transaction := ru.receivedTransaction.Transaction
+	streamId, err := shared.StreamIdFromBytes(transaction.StreamId)
+	if err != nil {
+		return nil, err
+	}
+	if transaction == nil {
+		return nil, RiverError(Err_INVALID_ARGUMENT, "transaction is nil")
+	}
+
+	switch ru.receivedTransaction.Kind {
+	case ReceivedBlockchainTransactionKind_RECEIVED_BLOCKCHAIN_TRANSACTION_KIND_UNSPECIFIED:
+		return nil, RiverError(Err_INVALID_ARGUMENT, "transaction kind is unspecified")
+	case ReceivedBlockchainTransactionKind_RECEIVED_BLOCKCHAIN_TRANSACTION_KIND_TIP:
+		// forward the tip to the channel stream of the fromUserAddress
```

Comment:
comment nit

`// forward the tip to the channel stream, attributed to fromUserAddress` maybe?

Reply:
`// forward the tip to the stream as a member event, preserving the original sender as the from address`